components/brave_vpn/browser/connection/wireguard/win/wireguard_utils.cc

```diff
@@ -0,0 +1,211 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_vpn/browser/connection/wireguard/win/wireguard_utils.h"
+
+#include <objbase.h>
+#include <stdint.h>
+#include <wrl/client.h>
+#include <utility>
+
+#include "base/base64.h"
+#include "base/logging.h"
+#include "base/strings/string_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/single_thread_task_runner.h"
+#include "base/task/thread_pool.h"
+#include "base/win/com_init_util.h"
+#include "base/win/scoped_bstr.h"
+#include "brave/components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/brave_wireguard_manager_idl.h"
+#include "brave/components/brave_vpn/browser/connection/wireguard/win/brave_vpn_wireguard_service/service_constants.h"
+
+namespace brave_vpn {
+
+namespace {
+// Template for wireguard config generation.
+constexpr char kWireguardConfigTemplate[] = R"(
+  [Interface]
+  PrivateKey = {client_private_key}
+  Address = {mapped_ipv4_address}
+  DNS = {dns_servers}
+  [Peer]
+  PublicKey = {server_public_key}
+  AllowedIPs = 0.0.0.0/0, ::/0
+  Endpoint = {vpn_server_hostname}:51821
+)";
+
+}  // namespace
+
+namespace wireguard {
+
+absl::optional<std::string> CreateWireguardConfig(
+    const std::string& client_private_key,
+    const std::string& server_public_key,
+    const std::string& vpn_server_hostname,
+    const std::string& mapped_ipv4_address,
+    const std::string& dns_servers) {
+  if (client_private_key.empty() || server_public_key.empty() ||
+      vpn_server_hostname.empty() || mapped_ipv4_address.empty() ||
+      dns_servers.empty()) {
+    return absl::nullopt;
+  }
+  std::string config = kWireguardConfigTemplate;
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{client_private_key}",
+                                     client_private_key);
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{server_public_key}",
+                                     server_public_key);
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{vpn_server_hostname}",
+                                     vpn_server_hostname);
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{mapped_ipv4_address}",
+                                     mapped_ipv4_address);
+  base::ReplaceSubstringsAfterOffset(&config, 0, "{dns_servers}", dns_servers);
+  return config;
+}
+
+bool EnableBraveVpnWireguardServiceImpl(const std::string& config) {
+  base::win::AssertComInitialized();
+  HRESULT hr = S_OK;
+  Microsoft::WRL::ComPtr<IBraveVpnWireguardManager> service;
+  CoCreateInstance(brave_vpn::GetBraveVpnWireguardServiceClsid(), nullptr,
+                   CLSCTX_LOCAL_SERVER,
+                   brave_vpn::GetBraveVpnWireguardServiceIid(),
+                   IID_PPV_ARGS_Helper(&service));
+  if (FAILED(hr)) {
```

Comment:
This will never fail because the `hr` variable was last set to the `S_OK` constant. Perhaps it needs to be set to the return value of `CoCreateInstance()`?

Reply:
done